### ISSUE TITLE
[Deliverable] Waku Sync

### ISSUE THREAD

**Issue by chair28980** [waku-org/pm#132]
## Summary

- **Goal:** After this upgrade, the network will provide distributed and synchronised store services.
- **Deliverables**: (2), (3)
- **Research tracks:** Message Reliability
- **Estimated date of completion:** 2024Q2

An improved version of the Store protocol, marking a crucial increment towards a synchronisation protocol:

- introduces the concept of deterministic message hashes to index messages
- considers the Store as a key-value store
- allows for querying a list of keys (message hashes) from the Store
- allows for querying for the full message content (values) or a set of keys from the Store
- keeps all previous value-based filtering (e.g. content topic, timestamp) in place

The [proposed PR](https://github.com/waku-org/waku-proto/pull/10/files) to simplify the Store protocol and use message
hashes as index/cursor, can be used as a starting point.

### Store v3 - store synchronisation

- **Goal:** Upgrade the Store service capability in the network from a collection of local, unsynchronised,
semi-centralised (trusted) service nodes to a decentralised service capability in the network with inter-node synchronisation.
- **Deliverables**: (2), (3)
- **Research tracks:** Message Reliability
- **Estimated date of completion:** 2024Q2

Building on Store v3-beta, this version of Store includes basic synchronisation between nodes.
This will probably include:

- a protocol/heuristic to resume store services after an offline period
- a protocol/heuristic to periodically compare local key-value store with other nodes and find missing keys
- a protocol/heuristic to periodically download the messages (values) for missing keys from other store nodes

Note that this can either be

- (i) designing a new synchronisation protocol built into the Store protocol
- (ii) adapting existing synchronisation building blocks (e.g. a [Prolly Tree library](https://docs.canvas.xyz/blog/2023-05-04-merklizing-the-key-value-store.html)) into Store, or
- (iii) simply swapping the key-value store itself to a synchronised backend, such as [GossipLog](https://github.com/canvasxyz/canvas/tree/c1c208032b56d6b474996cb7da7bda5c55b69463/packages/gossiplog)

IMO (iii) should be pursued as the preferred option, as far as possible.

**Comment by SionoiS** [comment]
- RFC and specs.
    - List the assumptions we are making 
    - Define the protobuf messages
    - Refer to negentropy for how it works

- New Nwaku protocol
    - This can be done without the bindings for the negentropy lib, just mock the payloads.
    - Code the interactions between 2 peers.
    - Define how other parts of the code will interact with Waku Sync

- Nim bindings for negentropy
    - Should be easy to do since we are exchanging mostly bytes through the language bridge.
    - Nim will not manage memory in this case. All the hashes will be sent across since storage will be on the C++ side.

- Hooking negentropy to a DB
    - They use LMDB we don't. How hard will it be to switch to postgres is unknown.
    - Could we switch to LMDB instead of postgres? Or maybe use both?

Tasks are mostly independent of each other and can be tackled in any order.

**Comment by SionoiS** [comment]
@chair28980 work has been scoped for research and nwaku. Not sure about the other teams.

**Comment by Ivansete-status** [comment]
Great summary @SionoiS !

May elaborate more this point plz?

> Nwaku: Nim bindings for negentropy

---
Re the following, we will need to switch to another solution that is more convenient for Waku and therefore I see the green light for investigating other databases. _Postgres_ is aimed to work for _Status_ and we can maintain both for now. 
> Could we switch to LMDB instead of postgres? Or maybe use both?

**Comment by chair28980** [comment]
Assigning @jm-clius as owner of the Deliverable and @SionoiS as owner of the owner of the two open epics.

**Comment by jm-clius** [comment]
@fryorcraken I agree. @AYAHASSAN287 has also confirmed today that QA is coming to an end this week and was successful.

@SionoiS could you double check task lists and close this and related issues as completed?

**Comment by Ivansete-status** [comment]
Considering the rule of "every deliverable should be dogfooded", shouldn't we wait to close that until the dogfooding is completed successfully?
@SionoiS correct me if I'm wrong and maybe I overlooked it somehow

**Comment by SionoiS** [comment]
I could leave it open but I would argue that these things are never "completed"... 

> "every deliverable should be dogfooded"

What kind of dogfooding we talking about here?

**Comment by fryorcraken** [comment]
dogfooding meaning that we have spent time testing out the feature in a live environment. 
Usually it means enabled in waku sandbox fleet, nwaku-compose, and potential building quick apps to demonstrate.

I agree with @Ivansete-status here because it seems that Waku sync is not even enabled in `nwaku-compose`...
